### ISSUE TITLE
longjmp could not return to the right place where setjmp is called

### ISSUE THREAD

**Issue by xxuejie** [emscripten-core/emscripten#1087]
I found yet another bug in asm.js version of setjmp/longjmp when testing webruby. The problem is that when a longjmp is called, the program could not return to the correct place where setjmp is called.

I'm sorry to say I still didn't find a way to create a minimal version of the bug. With O2 optimization it feels hard for me to pinpoint the correct place, whenever I tried to remove something from the program, the bug would disappear.

But I still manage to create two version of the bytecodes at [here](https://www.box.com/s/jvjdy53yrofrbfmvc1iq). The `failure` folder contains the bytecodes with the bug, while the `success` folder contains the correct bytecodes without any problems. The different between these two version is documented here:

The bug occurs in a function called `mrb_exc_raise`, when mruby is running a test case called `Integer#chr`(actually the bug has nothing to do with this test case, by removing some irrelevant test cases we could make it pass. The problem has to do with the stack or other environments when longjmp is called in this test). Normally, the code for this function is like this:

``` c
void
mrb_exc_raise(mrb_state *mrb, mrb_value exc)
{
  mrb->exc = mrb_obj_ptr(exc);
  fprintf(stderr, "mrb_exc_raise exc %p!\n", mrb->exc);
  exc_debug_info(mrb, mrb->exc);
  if (!mrb->jmp) {
    mrb_p(mrb, exc);
    abort();
  }
  fprintf(stderr, "Calling longjmp in exc_raise!\n");
  longjmp(*(jmp_buf*)mrb->jmp, 1);
}

```

The two `fprintf` functions are actually added by me to pinpoint this location. However, if we add a debug-print function here, and make the function look like this:

``` c
void
mrb_exc_raise(mrb_state *mrb, mrb_value exc)
{
  mrb->exc = mrb_obj_ptr(exc);
  fprintf(stderr, "mrb_exc_raise exc %p!\n", mrb->exc);
  exc_debug_info(mrb, mrb->exc);
  if (!mrb->jmp) {
    mrb_p(mrb, exc);
    abort();
  }

  // This is the added line!
  mrb_p(mrb, exc);

  fprintf(stderr, "Calling longjmp in exc_raise!\n");
  longjmp(*(jmp_buf*)mrb->jmp, 1);
}
```

The bug would be gone. Everything works just fine. The added `mrb_p` function just take a look at what is in `exc` and print the information to console. So I guess the problem has to do with different optimizations used here between the two versions.

In a nutshell, the `failure` folder at [here](https://www.box.com/s/jvjdy53yrofrbfmvc1iq) contains the bytecode without calling `mrb_p` before `longjmp`, hence the longjmp could not work. The `success` folder contains the bytecode with calling `mrb_`, and the longjmp could work correctly.

Sorry for the irrelevant parts in this bug. I really wish I could make it simpler. Thanks for your help!
